### PR TITLE
feat: general wrapper framework

### PR DESCRIPTION
**Pull request by sash-a**
## What?
General observation and example rware wrapper which makes it easy to support more jumanji envs in the future :eyes: 
## Why?
To support more jumanji envs
## How?
Create an observation type with known fields. Wrap all new envs so that they return this observation type instead of their own custom observation.

### CODE REVIEW

File: mava/systems/rec_ippo_rware.py [hunk 1]
```diff
@@ -48,10 +48,11 @@
     RecActorApply,
     RecCriticApply,
     RNNLearnerState,
+    RnnObservation,
```
Comment (OmaymaMahjoub) [inline mava/systems/rec_ippo_rware.py:51]: just for consistency can you name it RNNObservation or RecObservation 😅

File: mava/types.py [hunk 2]
```diff
@@ -21,17 +21,58 @@
...
-Observation: TypeAlias = Any
+
+
+class Observation(NamedTuple):
+    agents_view: chex.Array
+    action_mask: chex.Array
+    step_count: chex.Numeric
+
+
+class ObservationGlobalState(NamedTuple):
```
Comment (sash-a) [inline mava/types.py:44]: I really like that you moved this into here :+1:

File: mava/types.py [hunk 3]
```diff
@@ -21,17 +21,58 @@
...
+    env_state: State
+    episode_returns: chex.Numeric
+    episode_lengths: chex.Numeric
+    # Information about the episode return and length for logging purposes.
+    episode_return_info: chex.Numeric
+    episode_length_info: chex.Numeric
+
+
 RnnObservation: TypeAlias = Tuple[Observation, Done]
+RnnGlobalObservation: TypeAlias = Tuple[ObservationGlobalState, Done]
```
Comment (sash-a) [inline mava/types.py:75]: :fire:

File: mava/wrappers/jumanji.py [hunk 4]
```diff
@@ -169,28 +136,34 @@ def observation_spec(self) -> specs.Spec[Observation]:
...
+    def modify_timestep(self, timestep: TimeStep) -> TimeStep[Observation]:
+        n_agents = self._env.num_agents
+        observation = Observation(
             agents_view=timestep.observation.agents_view,
             action_mask=timestep.observation.action_mask,
-            step_count=jnp.repeat(timestep.observation.step_count, self._env.num_agents),
+            step_count=jnp.repeat(timestep.observation.step_count, n_agents),
         )
-        return state, timestep
+        shared_reward = jnp.sum(timestep.reward)
```
Comment (sash-a) [inline mava/wrappers/jumanji.py:153]: Do we need this? I think separate rewards for each agent is a good thing allows them to learn independently

File: mava/types.py [hunk 5]
```diff
@@ -21,17 +21,58 @@
...
 Done: TypeAlias = chex.Array
 HiddenState: TypeAlias = chex.Array
-
-# Can't know the exact type of State or Timestep.
+# Can't know the exact type of State.
 State: TypeAlias = Any
-Observation: TypeAlias = Any
+
+
+class Observation(NamedTuple):
```
Comment (sash-a) [inline mava/types.py:38]: Could we take the docs from `ObservationGlobalState` and put them here and then only put the `global_state` doc in the observation global state.
```suggestion
class Observation(NamedTuple):
"""The observation that the agents see. Used for all decentralized systems.

    agents_view: the array that is passed to the networks during acting and learning.
    action_mask: boolean array specifying, for each agent, which action is legal.
    step_count: the number of steps elapsed since the beginning of the episode.
```

File: mava/types.py [hunk 6]
```diff
@@ -21,17 +21,58 @@
...
+    """The observation that the agent sees.
+    agents_view: the agents' view of other agents and shelves within their
+        sensor range. The number of features in the observation array
+        depends on the sensor range of the agent.
+    action_mask: boolean array specifying, for each agent, which action
+        (up, right, down, left) is legal.
+    global_state: the global state of the environment, which is the
+        concatenation of the agents' views.
+    step_count: the number of steps elapsed since the beginning of the episode.
+    """
```
Comment (sash-a) [inline mava/types.py:54]: Note you may have to run pre-commit after adding this as I'm not sure about line lenghts

```suggestion
    """The observation that the agents see. Used in centralized systems.
    
    Same as `Observation` except it adds a `global_state` attribute for centralized training.
  
    global_state: the global state of the environment. (Often is concatenation of the agents' views.)
    """
```

**Review by sash-a — COMMENTED** [review]
I can't approve :cry: but I am happy with this, just one question about the shared reward and a minor docstring thing

**Comment by sash-a** [comment]
Just for bookkeeping #951 was discovered in this PR, but we're deciding to fix it in a separate PR so we can get this one in ASAP.

File: mava/configs/system/ff_ippo.yaml [hunk 7]
```diff
@@ -1,6 +1,6 @@
 # --- Defaults FF-IPPO ---
 
-num_updates: 1000 # Number of updates
+num_updates: 10000 # Number of updates
```
Comment (OmaymaMahjoub) [inline mava/configs/system/ff_ippo.yaml:3]: can you reset it to 1000

File: mava/wrappers/jumanji.py (hunk 4 above)
Comment (OmaymaMahjoub) [reply mava/wrappers/jumanji.py:153]: I think we need to add it as a feature in the systems files instead of the wrapper, what do you think?

Comment (WiemKhlifi) [reply mava/wrappers/jumanji.py:153]: Oh it's a mistake I added that to test the LBF environment I'll remove it and add this in the LBF wrapper :+1:.
For the idea of using separate rewards or the shared in LBF, we can use a flag as Omayma mentioned that will be helpful in future to test networks with non-shared parameters :crossed_fingers:

File: mava/configs/system/ff_ippo.yaml (hunk 7 above)
Comment (WiemKhlifi) [reply mava/configs/system/ff_ippo.yaml:3]: Sure!

**Review by OmaymaMahjoub — CHANGES_REQUESTED** [review]
Can we add a feature called `use_team_reward` as we need it for a fair comparison between our work and existing works that rely on giving the agents the same reward measured by `mean(timestep.reward)`

**Comment by WiemKhlifi** [comment]
I made all the necessary changes in this PR #952, just to make this PR smaller.
> Can we add a feature called `use_team_reward` as we need it for a fair comparison between our work and existing works that rely on giving the agents the same reward measured by `mean(timestep.reward)`

File: mava/evaluator.py [hunk 8]
```diff
@@ -105,7 +105,9 @@ def evaluator_fn(trained_params: FrozenDict, rng: chex.PRNGKey) -> ExperimentOut
         # Add dimension to pmap over.
         step_rngs = jnp.stack(step_rngs).reshape(eval_batch, -1)
 
-        eval_state = EvalState(step_rngs, env_states, timesteps, 0, 0.0)
+        eval_state = EvalState(
+            step_rngs, env_states, timesteps, 0, jnp.zeros_like(timesteps.reward)
```
Comment (RuanJohn) [inline mava/evaluator.py:109]: Is this so that we can support team and individual rewards?

File: mava/systems/rec_ippo_rware.py [hunk 9]
```diff
@@ -89,7 +90,7 @@ class Actor(nn.Module):
     def __call__(
         self,
         policy_hidden_state: chex.Array,
-        observation_done: Tuple[chex.Array, chex.Array],
+        observation_done: RnnObservation,
```
Comment (RuanJohn) [inline mava/systems/rec_ippo_rware.py:93]: SImilarly here. We should please go for `RNN` then instead of `Rnn`

File: mava/systems/rec_mappo_rware.py [hunk 10]
```diff
@@ -42,18 +42,20 @@
     ExperimentOutput,
     HiddenStates,
     LearnerFn,
+    ObservationGlobalState,
     OptStates,
     Params,
     PPOTransition,
     RecActorApply,
     RecCriticApply,
+    RnnGlobalObservation,
```
Comment (RuanJohn) [inline mava/systems/rec_mappo_rware.py:51]: Same as above

File: mava/types.py [hunk 11]
```diff
@@ -21,17 +21,60 @@
...
-
-# Can't know the exact type of State or Timestep.
+# Can't know the exact type of State.
 State: TypeAlias = Any
-Observation: TypeAlias = Any
+
+
+class Observation(NamedTuple):
+    """The observation that the agent sees.
+    agents_view: the agents' view of other agents and shelves within their
```
Comment (RuanJohn) [inline mava/types.py:40]: Since we want this to be general the docstring should reflect this please. This docstring looks like it is rware specific.

File: mava/types.py [hunk 12]
```diff
@@ -21,17 +21,60 @@
...
+        sensor range. The number of features in the observation array
+        depends on the sensor range of the agent.
+    action_mask: boolean array specifying, for each agent, which action is legal.
+    step_count: the number of steps elapsed since the beginning of the episode.
+    """
+
+    agents_view: chex.Array
```
Comment (RuanJohn) [inline mava/types.py:47]: Also for consistency, can we please have shape comments here as well like below.

File: mava/types.py [hunk 13]
```diff
@@ -21,17 +21,60 @@
...
+    action_mask: chex.Array
+    step_count: chex.Numeric
+
+
+class ObservationGlobalState(NamedTuple):
+    """The observation seen by agents in centralized systems.
```
Comment (RuanJohn) [inline mava/types.py:53]:
```suggestion
    """The observation seen by agents in centralised systems.
```

File: mava/types.py [hunk 14]
```diff
@@ -21,17 +21,60 @@
...
+    Extends `Observation` by adding a `global_state` attribute for centralized training.
```
Comment (RuanJohn) [inline mava/types.py:54]:
```suggestion
    Extends `Observation` by adding a `global_state` attribute for centralised training.
```

File: mava/wrappers/jumanji.py [hunk 15]
```diff
@@ -93,6 +60,46 @@ def step(
         return state, timestep
 
 
+class RwareWrapper(Wrapper):
```
Comment (RuanJohn) [inline mava/wrappers/jumanji.py:63]: I really like this!!

File: mava/wrappers/jumanji.py [hunk 16]
```diff
@@ -169,102 +178,52 @@ def observation_spec(self) -> specs.Spec[Observation]:
...
-            [0] * self._env.num_agents,
-            [self._env.time_limit] * self._env.num_agents,
-            "step_count",
-        )
-        return self._env.observation_spec().replace(step_count=step_count)
-
-
-class RwareMultiAgentWithGlobalStateWrapper(Wrapper):
-    """Multi-agent wrapper for the Robotic Warehouse environment.
+class GlobalStateWrapper(Wrapper):
```
Comment (RuanJohn) [inline mava/wrappers/jumanji.py:181]: I wonder if we shouldn't have a `wrapper/mava.py` file that includes the more general wrappers like the agent IDs and the global state wrappers? Since I foresee us reusing these for any env we wrap.

**Review by RuanJohn — CHANGES_REQUESTED** [review]
Thank you for all the work! Super nice changes! Just some minor comments and suggestions here and there.

File: mava/systems/rec_ippo_rware.py (hunk 1 above)
Comment (WiemKhlifi) [reply mava/systems/rec_ippo_rware.py:51]: I think RecObservation may be better if you agree on this I'll change it 👌

File: mava/evaluator.py (hunk 8 above)
Comment (WiemKhlifi) [reply mava/evaluator.py:109]: Actually, this was changed because we removed the reshaping of the reward outside of the system here [https://github.com/instadeepai/Mava/blob/7e500aaa2c657ce818689c589020f39284391da4/mava/systems/ff_ippo_rware.py#L152] and added it the wrapper in the modify_timesteps [https://github.com/instadeepai/Mava/blob/7e500aaa2c657ce818689c589020f39284391da4/mava/wrappers/jumanji.py#L77C6-L77C6] so that in future we can use the same system.

File: mava/types.py (hunk 11 above)
Comment (WiemKhlifi) [reply mava/types.py:40]: I missed it thank you Ruan I changed it! 🙏

File: mava/types.py (hunk 12 above)
Comment (WiemKhlifi) [reply mava/types.py:47]: Sure!

File: mava/wrappers/jumanji.py (hunk 16 above)
Comment (WiemKhlifi) [reply mava/wrappers/jumanji.py:181]: I really like this idea 🔥 ! We can name the file mava.py (like jumanji.py, gym.py ...) and then we change and adjust the environments wrapper separately 🙌

File: mava/types.py [hunk 17]
```diff
@@ -21,17 +21,60 @@
...
-# Can't know the exact type of State or Timestep.
+# Can't know the exact type of State.
 State: TypeAlias = Any
-Observation: TypeAlias = Any
-RnnObservation: TypeAlias = Tuple[Observation, Done]
+
+
+class Observation(NamedTuple):
+    """The observation that the agent sees.
+    agents_view: the agents' view of other agents and items within their
```
Comment (RuanJohn) [inline mava/types.py:40]: Maybe a simpler docstring here is to just say `the agent's view of the environment.`

File: mava/wrappers/jumanji.py (hunk 16 above)
Comment (RuanJohn) [reply mava/wrappers/jumanji.py:181]: @sash-a, keen to hear your thoughts here? I think we should everything jumanji specific in this jumanji file and move general wrappers to another file. Calling the file `mava.py` is a bit misleading since Mava isn't an environment. My suggestion is to call it `shared.py` since it will be wrappers shared between envs.

File: mava/types.py (hunk 17 above)
Comment (WiemKhlifi) [reply mava/types.py:40]: Thanks for suggesting this I'll change it 🙌

**Review by RuanJohn — APPROVED** [review]
Thank you @WiemKhlifi and @sash-a :fire:

File: mava/wrappers/jumanji.py (hunk 16 above)
Comment (RuanJohn) [reply mava/wrappers/jumanji.py]: I think that is okay, since the base Mava API is an extension of the Jumanji API. Thank you for the changes :raised_hands:

**Review by OmaymaMahjoub — APPROVED** [review]
Thanks @WiemKhlifi  and @sash-a  :fire:
